## Imports

In [1]:
%load_ext autoreload
%autoreload 2

import logging

import matplotlib.pyplot as plt
import seaborn as sns

import L05_01_learning_theory_02_bin_analogy_ml_utils as utils

# Set plotting style.
sns.set_style("whitegrid")
plt.rcParams["figure.figsize"] = (12, 6)

WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.


In [2]:
import msml610_utils as ut

ut.config_notebook()

# Initialize logger.
logging.basicConfig(level=logging.INFO)
_LOG = logging.getLogger(__name__)

vim support installed: restart the notebook, if needed


Python 3.12.3
Linux 589569fe8102 6.12.67-linuxkit #1 SMP Sun Jan 25 02:26:28 UTC 2026 aarch64 aarch64 aarch64 GNU/Linux


## Cell 1: Visual Bin: Population of Marbles.

- This visualization shows a bin filled with red and green marbles representing the unknown population
- The parameter $\mu$ represents the true proportion of red marbles in the bin

In [3]:
utils.cell1_draw_bin_with_marbles_interactive()

## Cell 2: Single Experiment: Is $\nu$ Close to $\mu$?

- This cell demonstrates a single sampling experiment from the bin.
- We draw $N$ marbles randomly (with replacement) and compute the sample proportion $\nu$.
- The key question is: How close is $\nu$ to the true $\mu$?

In [4]:
utils.cell2_plot_single_experiment_interactive()

## Cell 2.1: Limitations of Single Experiments

- Single experiments don't tell the full story
  - We need to know $P(|\nu - \mu| > \epsilon)$
- Questions to consider:
  - What if we got unlucky in our sample?
  - How confident can we be that $\nu \approx \mu$?
  - Does sample size $N$ matter? How much?
  - "Let's repeat this many times..."

## Cell 3: Monte Carlo Simulation: Distribution of $\nu$

- Run many sampling experiments (n_experiments times)
- Each experiment: draw $N$ marbles, compute $\nu$
- Collect all $\nu$ values and visualize their distribution
- Key questions:
  - How are the $\nu$ values distributed?
  - What fraction of experiments have $|\nu - \mu| > \epsilon$?
  - Does the distribution concentrate around $\mu$?

In [5]:
utils.cell3_monte_carlo_simulation_interactive()

## Cell 3.1: Observations from Monte Carlo Simulation

- The distribution of $\nu$ values clusters around the true $\mu$
- As $N$ increases, the distribution becomes tighter (smaller variance)
- The empirical probability $P(|\nu - \mu| > \epsilon)$ decreases with larger $N$
- This demonstrates the Law of Large Numbers in action